more_itertools/recipes.py

```diff
@@ -1075,3 +1077,89 @@ def totient(n):
     for prime in set(factor(n)):
         n -= n // prime
     return n
+
+
+# Miller–Rabin primality test: Testing against small sets of bases
+# https://w.wiki/CGwr
+_perfect_tests = [
+    (2047, (2,)),
+    (9080191, (31, 73)),
+    (4_759_123_141, (2, 7, 61)),
+    (1122004669633, (2, 13, 23, 1662803)),
+    (2152302898747, (2, 3, 5, 7, 11)),
+    (3474749660383, (2, 3, 5, 7, 11, 13)),
+    (341550071728321, (2, 3, 5, 7, 11, 13, 17)),
+    (3825123056546413051, (2, 3, 5, 7, 11, 13, 17, 19, 23)),
+    (318665857834031151167461, (2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37)),
+    (
+        3317044064679887385961981,
+        (2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41),
+    ),
+]
+
+
+@lru_cache
+def _shift_to_odd(n):
+    'Return s, d such that 2**s * d == n'
+    s = ((n - 1) ^ n).bit_length() - 1
+    d = n >> s
+    assert (1 << s) * d == n and d & 1
+    return s, d
+
+
+def _strong_probable_prime(n, base):
+    assert n > 2 and n & 1
+    s, d = _shift_to_odd(n - 1)
```

Comment:
Also add another precondition:

`assert 2 <= base < n`

This will detect cases where the `_perfect_tests` table is misordered. For example, if the `(2047, (2,))` entry is omitted, `is_prime(31)` would give an incorrect result.